app/services/data_api/tad_subjects_export.rb

```diff
@@ -0,0 +1,177 @@
+          indexed_counts[key] ||= {
+            adjusted_applications: 0,
+            adjusted_offers: 0,
+            pending_conditions: 0,
+            recruited: 0,
+          }
+
+          increment_counts(indexed_counts, key, status, subjects.count)
+        end
+      end
+      roundup_results(flatten_results(indexed_counts))
+    end
+
+    def subjects_indexed_by_status(result)
+      subjects_indexed_by_status = {}
+      result['subject_statuses'].each do |subject, status|
+        subjects_indexed_by_status[status.to_sym] ||= []
+        subjects_indexed_by_status[status.to_sym] << subject
+      end
+      subjects_indexed_by_status
+    end
+
+    def subjects_for_most_progressed_statuses(subjects_indexed_by_status)
+      if subjects_indexed_by_status.key?(:recruited)
+        subjects_indexed_by_status.select! { |status| status == :recruited }
+      elsif subjects_indexed_by_status.key?(:pending_conditions)
+        subjects_indexed_by_status.select! { |status| status == :pending_conditions }
+      elsif subjects_indexed_by_status.key?(:offer)
+        subjects_indexed_by_status.select! { |status| status == :offer }
+      else
+        subjects_indexed_by_status = { other: subjects_indexed_by_status.values.flatten }
+      end
+      status, subjects = subjects_indexed_by_status.to_a.first
+
+      [status, subjects]
+    end
+
+    def increment_counts(indexed_counts, key, status, count)
+      case status
+      when :recruited
+        indexed_counts[key][:recruited] += 1
+      when :pending_conditions
+        indexed_counts[key][:pending_conditions] += 1
+      when :offer
+        indexed_counts[key][:adjusted_offers] += (1 / count.to_f)
+      else
+        indexed_counts[key][:adjusted_applications] += (1 / count.to_f)
+      end
+    end
+
+    def flatten_results(counts)
+      counts.map do |key, count|
+        {
+          subject: key[0],
+          candidate_domicile: key[1],
+          candidate_nationality: key[2],
+        }.merge(count)
+      end
+    end
+
+    def roundup_results(counts)
+      counts.map do |count|
+        count.merge(
+          recruited: count[:recruited].ceil,
+          pending_conditions: count[:pending_conditions].ceil,
+          adjusted_offers: count[:adjusted_offers].ceil,
+          adjusted_applications: count[:adjusted_applications].ceil,
+        )
+      end
+    end
+
+    def type_map
+      @type_map ||= PG::BasicTypeMapForResults.new(ActiveRecord::Base.connection.raw_connection)
+    end
+
+    def query
+      <<-SQL
+        SELECT
+          application_forms.candidate_id,
+          application_forms.id,
+          ARRAY_AGG(ARRAY[subjects.name, application_choices.status]) subject_statuses,
+          CASE
+            WHEN first_nationality IS NULL OR first_nationality = '' THEN NULL
+            WHEN array[#{uk_nationalities}]::varchar[] && array[first_nationality, second_nationality, third_nationality, fourth_nationality] THEN 'UK'
```

Comment:
Just checking my understanding of this. It essentially grabs the nationalities and tries to match them to the array of nationalities right?

i.e in ruby it's doing
`([uk_nationalties] & [candidates.nationalties ]).any?`

Reply:
@davidgisbey yes it's logically the same as that Ruby snippet. I can never remember all the array operators TBH https://www.postgresql.org/docs/12/functions-array.html